### ISSUE TITLE
Serialized shadow blocks can become invalid

### ISSUE THREAD

**Issue by NeilFraser** [google/blockly#4883]
Open the playground and load this XML:

```
<xml xmlns="https://developers.google.com/blockly/xml">
  <block type="controls_repeat" x="112" y="112">
    <statement name="DO">
      <shadow type="procedures_callnoreturn">
        <mutation name="func1"></mutation>
      </shadow>
    </statement>
  </block>
  <block type="procedures_defnoreturn" x="362" y="138">
    <field name="NAME">func1</field>
  </block>
  <block type="text_print" x="113" y="237">
  </block>
</xml>
```

Drag the print block to clobber the shadowed function call block.
Rename the function definition from 'func1' to 'func2'.
Drag the print block away from the loop.

The result is that the function call was not renamed since it was serialized as a shadow.  The program is now invalid.

Note that if the print block is deleted, the invalid func1 shadow block also gets deleted.  This is the only case I've seen where a shadow block can be deleted.

Should clobbered shadow blocks remain as real Block objects returned by workspace.getAllBlocks so that they can be affected by renames?  Or would this create even more edge cases?

**Comment by BeksOmega** [comment]
I think having `getAllBlocks` return shadows would cause similar problems to the ones that insertion markers caused (eg #2236), so I wouldn't want to touch it personally haha.

But I'm wondering, is there a use case for having procedure call blocks as shadows (or a similar use case where a block reacts to changes in the workspace)? It seems weird to provide a call block as a default when the user could delete the function definition and make the call undefined. I imagine if an application wanted to do something similar, they would use a custom not-deletable hat block with a custom call block or something.

**Comment by NeilFraser** [comment]
Agreed, this smells exactly like the pain we went through with insertion markers.

No, there's no known use case for this abomination.  It was discovered during a discussion with Chris.

That said, clean solution don't have sharp edges lurking in the dark.  So it's worth thinking about what an acceptable solution would be.  It doesn't need to be user-friendly (since this is apparently only theoretical), but it should keep the system in a valid state.

**Comment by cpcallen** [comment]
> It was discovered during a discussion with Chris.

Shh!  If you keep telling people about all the weird theoretical edge cases I uncover they'll stop inviting me to their ~~parties~~projects.  😭
